web/src/test/auto/headless/app/browser/hardware-event-processing.tests.ts

```diff
@@ -0,0 +1,590 @@
+          LisVirtualKey: true,
+          device: DUMMY_DEVICE,
+          kName: "K_A"
+        });
+
+        const finalKeyEvent = processForMnemonicsAndLegacy(
+          baseKeyEvent,
+          new Keyboard({KM:1}),
+          'us'
+        );
+
+        // May be the same instance, but the instance gets mutated.  Compare against literals
+        // for expected values.
+        assert.equal(finalKeyEvent.Lcode, 'a'.charCodeAt(0));
+        assert.equal(finalKeyEvent.Lmodifiers, 0);
+        assert.isTrue(finalKeyEvent.LisVirtualKey);
+      });
+    });
+
+    describe('legacy keyboards', () => {
+
+      it("simple shifted 'a'", () => {
+        const baseKeyEvent = new KeyEvent({
+          Lcode: KeyCodes.K_A,
+          Lmodifiers: ModifierCodes.SHIFT,
+          Lstates: ModifierCodes.NO_CAPS | ModifierCodes.NO_NUM_LOCK | ModifierCodes.NO_SCROLL_LOCK,
+          LisVirtualKey: true,
+          device: DUMMY_DEVICE,
+          kName: "K_A"
+        });
+
+        const finalKeyEvent = processForMnemonicsAndLegacy(
+          baseKeyEvent,
+          new Keyboard({KM:undefined}),
+          'us'
+        );
+
+        // The event object gets wholesale-replaced.
+        assert.notStrictEqual(finalKeyEvent, baseKeyEvent);
+
+        assert.isFalse(finalKeyEvent.LisVirtualKey);
+        assert.equal(finalKeyEvent.Lcode, 'A'.charCodeAt(0));
+        assert.equal(finalKeyEvent.Lmodifiers, 0);
+      });
+
+      it("simple unshifted 'a'", () => {
+        const baseKeyEvent = new KeyEvent({
+          Lcode: KeyCodes.K_A,
+          Lmodifiers: 0,
+          Lstates: ModifierCodes.NO_CAPS | ModifierCodes.NO_NUM_LOCK | ModifierCodes.NO_SCROLL_LOCK,
+          LisVirtualKey: true,
+          device: DUMMY_DEVICE,
+          kName: "K_A"
+        });
+
+        const finalKeyEvent = processForMnemonicsAndLegacy(
+          baseKeyEvent,
+          new Keyboard({KM:undefined}),
+          'us'
+        );
+
+        // The event object gets wholesale-replaced.
+        assert.notStrictEqual(finalKeyEvent, baseKeyEvent);
+
+        assert.isFalse(finalKeyEvent.LisVirtualKey);
+        assert.equal(finalKeyEvent.Lcode, 'a'.charCodeAt(0));
+        assert.equal(finalKeyEvent.Lmodifiers, 0);
+      });
+
+      it("'a' during alt", () => {
+        const baseKeyEvent = new KeyEvent({
+          Lcode: KeyCodes.K_A,
+          Lmodifiers: ModifierCodes.ALT,
+          Lstates: ModifierCodes.NO_CAPS | ModifierCodes.NO_NUM_LOCK | ModifierCodes.NO_SCROLL_LOCK,
+          LisVirtualKey: true,
+          device: DUMMY_DEVICE,
+          kName: "K_A"
+        });
+
+        const finalKeyEvent = processForMnemonicsAndLegacy(
+          baseKeyEvent,
+          new Keyboard({KM:undefined}),
+          'us'
+        );
+
+        // Legacy keyboards expect this to be false; this acts as a filter for legacy keyboards
+        // to prevent key processing with held CTRL / ALT.
+        assert.isTrue(finalKeyEvent.LisVirtualKey);
```

Comment:
I don't understand the comment. It speaks about "expect this to be false" but then tests `assert.isTrue` !?

Reply:
Based on the changes in #10437, we assume that KMW 1.0 keyboards were only ever supposed to handle `default` and `shift` layers.  They also were not aware of the `LisVirtualKey` field.  

If that field is set to true, all of a KMW-1.0 keyboard's rules will fail to match as a result.  So, if it's `true` for events with modifiers it's not supposed to handle, _that_ is the mechanism that prevents such a keyboard from doing anything on layers other than `default` and `shift`.

As this unit test is an `alt` layer, we want the mismatch here.

Reply:
Will replace with

```typescript
// KMW 1.0 keyboards expect this to be false; setting it to true prevents the keyboard
// from matching the event to a rule.  As KMW 1.0 keyboards shouldn't handle modifiers
// other than shift, this acts as a filter to prevent rule processing for the unsupported
// modifier.
```